### PR TITLE
[FIX] #145 - when using TLS reconnects would fail because of listener leaks (events firing on leaked streams)

### PR DESCRIPTION
**Pull request by aricart**
[FIX] node client fired error on any protocol error. Some protocol errors such as 'stale connection' should have initiated a reconnect.
[FIX] if a client attempted to publish/subscribe to a subject for which it had no permission, an error was generated. New event ('permission_error') will now be sent to the client.

@blazedd effectively submitted a [similar PR](https://github.com/nats-io/node-nats/pull/146).

### CODE REVIEW

**Comment by aricart** [comment]
Note that I have included 'authorization timeout' as a non-fatal error. Other clients don't handle this, we can remove this behaviour or enhance other clients to reconnect if a server fails to authorize because of some internal load issue. Not sure if this is the right thing to do.

**Comment by derekcollison** [comment]
The server should be expected to disconnect after an authorization timeout. I would cleanup the connection and fire the event callbacks.

**Comment by aricart** [comment]
Fire 'error'?

**Comment by derekcollison** [comment]
Yes I believe so..

**Comment by aricart** [comment]
Updated so the client disconnects on the timeout error. Added another test without tls which will always get the error from the server.

File: lib/nats.js [hunk 1]
```diff
@@ -91,6 +91,9 @@ var VERSION = '0.7.18',
     NATS_PROTOCOL_ERR = 'NATS_PROTOCOL_ERR',
     REQ_TIMEOUT = 'REQ_TIMEOUT',
     REQ_TIMEOUT_MSG_PREFIX = 'The request timed out for subscription id: ',
+    STALE_CONNECTION_ERR = "stale connection",
+    PERMISSIONS_ERR = "permissions violation",
+    AUTHORIZATION_TIMEOUT_ERR = "authorization timeout",
```
Comment (kozlovic) [inline lib/nats.js]: Not used.

File: test/perms.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        ncu.writeFile(cf, ncu.JsonToYaml(conf));
+        server = nsc.start_server(PORT, ['-c', cf], done);
+    });
+
+    // Shutdown our server
+    after(function () {
+        nsc.stop_server(server);
+    });
+
+    it('foo can subscribe/pub foo', function (done) {
```
Comment (kozlovic) [inline test/perms.js]: I would not test the positive cases here, this is a server test, not a client. You want to test a part of the code that handles the failure, so I would keep only negative tests where we try to pub/sub and get an error and check that connection is not closed.

File: lib/nats.js (hunk 1 above)
Comment (aricart) [inline lib/nats.js]: good catch

**Comment by kozlovic** [comment]
LGTM

**Pull request merged**